polaris-core/src/main/java/org/apache/polaris/core/persistence/IntegrationPersistence.java

```diff
@@ -0,0 +1,129 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.persistence;
+
+import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.entity.PolarisBaseEntity;
+import org.apache.polaris.core.entity.PolarisPrincipalSecrets;
+import org.apache.polaris.core.storage.PolarisStorageConfigurationInfo;
+import org.apache.polaris.core.storage.PolarisStorageIntegration;
+
+/**
+ * Interface for the necessary "peripheral integration" objects that are logically attached to core
+ * persistence entities but which typically involve additional separate external integrations
+ * related to identity/auth, kms/secrets storage, etc.
+ *
+ * <p>Implementations should orchestrate any necessary multi-phase protocols such as leasing an
+ * external resource before committing a reference to the external resource in the Polaris
+ * persistence layer, etc.
+ */
+public interface IntegrationPersistence {
+  /**
+   * Allows to retrieve to the secrets of a principal given its unique client id
+   *
+   * @param callCtx call context
+   * @param clientId principal client id
+   * @return the secrets
+   */
+  @Nullable
+  PolarisPrincipalSecrets loadPrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String clientId);
+
+  /**
+   * generate and store a client id and associated secrets for a newly created principal entity
+   *
+   * @param callCtx call context
+   * @param principalName name of the principal
+   * @param principalId principal id
+   */
+  @Nonnull
+  PolarisPrincipalSecrets generateNewPrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String principalName, long principalId);
+
+  /**
+   * Rotate the secrets of a principal entity, i.e. make the specified main secrets the secondary
+   * and generate a new main secret
+   *
+   * @param callCtx call context
+   * @param clientId principal client id
+   * @param principalId principal id
+   * @param reset true if the principal secrets should be disabled and replaced with a one-time
+   *     password
+   * @param oldSecretHash the principal secret's old main secret hash
+   */
+  @Nullable
+  PolarisPrincipalSecrets rotatePrincipalSecrets(
```

Comment:
This feels too much for a persistence API. Why should persistence care about how secrets are rotated?

Reply:
Right now it's the semantics of how rotate interacts with reset and "require rotate" that impacts the behavior of "unlocking" principals to be usable. This might be able to only be done in the PolarisMetaStoreManager layer though. We need to stash these methods somewhere while we try to figure out how to pull them into a different layer.

Reply:
Need to think about external IdPs as well.

Reply:
ok, let's defer this for now